pkg/placement/declarations.go

```diff
@@ -0,0 +1,387 @@
+// all the objects in that namespace are included.
+//
+// A workload may include objects of kinds that are built into
+// the edge cluster.  By built-in we mean that these kinds are both
+// already known (regardless of whether it is via being built into
+// the apiserver or added to it by either form of aggregation) and
+// not managed by edge workload management.
+// It is the user's responsibility to make the "what" predicate
+// match the corresponding CRD when the workload includes an object
+// of a kind that is not built into the edge cluster.
+type WorkloadParts map[WorkloadPartID]WorkloadPartDetails
+
+// WorkloadPartID identifies part of a workload.
+type WorkloadPartID struct {
+	APIGroup string
+
+	// Resource is the lowercase plural way of identifying the kind of object
+	Resource string
+
+	Name string
+}
+
+// WorkloadPartDetails provides additional details about how the WorkloadPart
+// is to be included.
+type WorkloadPartDetails struct {
+	// APIVersion is version (no group) that the source workspace prefers to serve.
+	APIVersion string
+
+	// IncludeNamespaceObject is only interesting for a Namespace part, and
+	// indicates whether to include the details of the Namespace object;
+	// the objects in the namespace are certainly included.
+	// For other parts, this field holds `false`.
+	IncludeNamespaceObject bool
+}
+
+type WorkloadPart struct {
+	WorkloadPartID
+	WorkloadPartDetails
+}
+
+// ProjectionMapProducer tells the consumers what to project,
+// organized into three levels.
+type ProjectionMapProducer DynamicMapProducer[ProjectionKey, *ProjectionPerCluster]
+
+// ProjectionKey identifies the topmost level of organization,
+// the combinatin of the destination and the API group and resource.
+type ProjectionKey struct {
+	metav1.GroupResource
+	Destination SinglePlacement
+}
+
+// ProjectionPerCluster is the second level of organization.
+// It identifies the API version to use currently and holds
+// the map provider that gets to the lowest level of organization.
+type ProjectionPerCluster struct {
+	// APIVersion is the version to read.  Just the version, no group included
+	APIVersion string
+
+	// PerSourceCluster drives awareness of the relevant logical clusters
+	// and the work to do for each.
+	// This provider (a) requires consumers to be comparable and (b) deduplicates
+	// additions of consumers.
+	PerSourceCluster DynamicMapProducer[logicalcluster.Name, ProjectionDetails]
+}
+
+// ProjectionDetails modulates projection
+type ProjectionDetails struct {
+
+	// For namespaced resoruces, Namespaces can optionally be non-nil to restrict
+	// the namespaces read from.
+	Namespaces *k8ssets.String
+
+	// For non-namespaced objects, Names can optionally be non-nil to restrict
+	// the objects handled.
+	Names *k8ssets.String
+}
+
+// SinglePlacement extends the API struct with the UID of the SyncTarget.
+type SinglePlacement struct {
+	edgeapi.SinglePlacement
+	SyncTargetUID apimachtypes.UID
```

Comment:
We may need to change this because now #198 is merged.

It is just a kind reminder in case it is omitted, because it is easier to be forgotten comparing to other explicit usage of `ExternalName`.

Reply:
Yes, I will rebase this PR.